KSQL 5.2.1 UDF load failing

When upgrading to CP 5.2.1, all UDF loads that previously worked in 5.1.2 are failing. So I went back to basics and compiled the `multiply` function from the Confluent documentation, and that failed as well.

Notes:
I am using gradle and the `shadowJar` plugin. This worked successfully in 5.1.2.

I have tested using both of the following dependency statements:
`compile 'io.confluent.ksql:ksql-udf:5.2.1'`
`compile 'io.confluent.ksql:ksql-udf:5.1.2'`
I wanted to see if the issue was only with the new dependency, or the new KSQL Server. It seems to be the new KSQL Server, as either dependency (`5.1.2` or `5.2.1`) fails with the same error.

The error I'm getting during the startup of the KSQL Server is below. I've also attached my uber JAR created by gradle. I had to zip it first for GitHub... but inside the ZIP is a JAR.

```
INFO Adding UDF name='multiply' from path=/etc/ksql-server/ext/functions-1.0.0.jar (io.confluent.ksql.function.UdfLoader:211)
WARN Failed to add UDF to the MetaStore. name=multiply method=public long Multiply.multiply(long,long) (io.confluent.ksql.function.UdfLoader:220)
io.confluent.ksql.util.KsqlException: Failed to compile udf wrapper class for public long Multiply.multiply(long,long)
at io.confluent.ksql.function.UdfCompiler.compile(UdfCompiler.java:96)
at io.confluent.ksql.function.UdfLoader.lambda$handleUdfAnnotation$8(UdfLoader.java:214)
at io.github.lukehutch.fastclasspathscanner.scanner.ScanSpec$9.lookForMatches(ScanSpec.java:1390)
at io.github.lukehutch.fastclasspathscanner.scanner.ScanSpec.callMatchProcessors(ScanSpec.java:696)
at io.github.lukehutch.fastclasspathscanner.FastClasspathScanner.scan(FastClasspathScanner.java:1606)
at io.github.lukehutch.fastclasspathscanner.FastClasspathScanner.scan(FastClasspathScanner.java:1678)
at io.github.lukehutch.fastclasspathscanner.FastClasspathScanner.scan(FastClasspathScanner.java:1704)
at io.confluent.ksql.function.UdfLoader.loadUdfs(UdfLoader.java:141)
at io.confluent.ksql.function.UdfLoader.lambda$load$2(UdfLoader.java:111)
at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
at java.util.Iterator.forEachRemaining(Iterator.java:116)
at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
at io.confluent.ksql.function.UdfLoader.load(UdfLoader.java:111)
at io.confluent.ksql.rest.server.KsqlRestApplication.buildApplication(KsqlRestApplication.java:333)
at io.confluent.ksql.rest.server.KsqlRestApplication.buildApplication(KsqlRestApplication.java:307)
at io.confluent.ksql.rest.server.KsqlServerMain.createExecutable(KsqlServerMain.java:85)
at io.confluent.ksql.rest.server.KsqlServerMain.main(KsqlServerMain.java:50)
Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
at java.util.ArrayList$SubList.rangeCheck(ArrayList.java:1225)
at java.util.ArrayList$SubList.get(ArrayList.java:1042)
at java.util.Collections$UnmodifiableList.get(Collections.java:1309)
at com.squareup.javapoet.ClassName.<init>(ClassName.java:52)
at com.squareup.javapoet.ClassName.<init>(ClassName.java:43)
at com.squareup.javapoet.ClassName.enclosingClassName(ClassName.java:76)
at com.squareup.javapoet.CodeWriter.lookupName(CodeWriter.java:358)
at com.squareup.javapoet.ClassName.emit(ClassName.java:237)
at com.squareup.javapoet.CodeWriter.emit(CodeWriter.java:253)
at com.squareup.javapoet.CodeBlock.toString(CodeBlock.java:95)
at io.confluent.ksql.function.UdfTemplate.generateCode(UdfTemplate.java:55)
at io.confluent.ksql.function.UdfCompiler.generateCode(UdfCompiler.java:251)
at io.confluent.ksql.function.UdfCompiler.compile(UdfCompiler.java:90)
... 25 more
```

[functions-1.0.0.jar.zip](https://github.com/confluentinc/ksql/files/3107801/functions-1.0.0.jar.zip)

--- a/ksql-engine/src/main/java/io/confluent/ksql/function/UdfCompiler.java
+++ b/ksql-engine/src/main/java/io/confluent/ksql/function/UdfCompiler.java
@@ -15,6 +15,7 @@
 
 package io.confluent.ksql.function;
 
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableSet;
 import io.confluent.ksql.function.udaf.TableUdaf;
 import io.confluent.ksql.function.udaf.Udaf;
@@ -82,7 +83,8 @@ public class UdfCompiler {
     this.metrics = Objects.requireNonNull(metrics, "metrics can't be null");
   }
 
-  UdfInvoker compile(final Method method, final ClassLoader loader) {
+  @VisibleForTesting
+  public UdfInvoker compile(final Method method, final ClassLoader loader) {
     try {
       final IScriptEvaluator scriptEvaluator = createScriptEvaluator(method,
           loader,

--- a/ksql-engine/src/main/java/io/confluent/ksql/function/UdfTemplate.java
+++ b/ksql-engine/src/main/java/io/confluent/ksql/function/UdfTemplate.java
@@ -49,8 +49,10 @@ public final class UdfTemplate {
         .mapToObj(i -> "arg" + i)
         .collect(Collectors.joining(", "));
 
-    code.addStatement("return (($T) $L).$L($L)",
-                      method.getDeclaringClass(), obj, method.getName(), args);
+    // use the canonical name instead of relying on $T because there is a bug
+    // in javapoet when trying to compile using the default package
+    code.addStatement("return (($L) $L).$L($L)",
+        method.getDeclaringClass().getCanonicalName(), obj, method.getName(), args);
 
     final String codeString = code.build().toString();
     LOG.trace("Generated code:\n" + codeString);

--- a/ksql-engine/src/main/java/io/confluent/ksql/function/UdfTemplate.java
+++ b/ksql-engine/src/main/java/io/confluent/ksql/function/UdfTemplate.java
@@ -49,8 +49,6 @@ public final class UdfTemplate {
         .mapToObj(i -> "arg" + i)
         .collect(Collectors.joining(", "));
 
-    // use the canonical name instead of relying on $T because there is a bug
-    // in javapoet when trying to compile using the default package
     code.addStatement("return (($L) $L).$L($L)",
         method.getDeclaringClass().getCanonicalName(), obj, method.getName(), args);
 

--- a/ksql-engine/src/test/java/TestUdfWithNoPackage.java
+++ b/ksql-engine/src/test/java/TestUdfWithNoPackage.java
@@ -19,14 +19,24 @@ import static org.hamcrest.CoreMatchers.is;
 import io.confluent.ksql.function.UdfCompiler;
 import io.confluent.ksql.function.UdfInvoker;
 import java.util.Optional;
+import org.junit.Before;
+import org.junit.BeforeClass;
 import org.junit.Test;
 
 public class TestUdfWithNoPackage {
 
   private final UdfCompiler udfCompiler = new UdfCompiler(Optional.empty());
 
+  @BeforeClass
+  public static void ensureTestHasNoPackage() {
+    // guard against someone accidentally refactoring and moving this test
+    // into a non-empty package
+    assertThat(TestUdfWithNoPackage.class.getPackage().getName(), is(""));
+  }
+
   @Test
   public void shouldCompileMethodsWithNoPackage() throws Exception {
+    // motivated by https://github.com/square/javapoet/pull/723
     final UdfInvoker udf = udfCompiler
         .compile(getClass().getMethod("udf"), this.getClass().getClassLoader());
 

--- a/ksql-engine/src/test/java/TestUdfWithNoPackage.java
+++ b/ksql-engine/src/test/java/TestUdfWithNoPackage.java
@@ -14,31 +14,34 @@
  */
 
 import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.nullValue;
 import static org.hamcrest.MatcherAssert.assertThat;
 
 import io.confluent.ksql.function.UdfCompiler;
 import io.confluent.ksql.function.UdfInvoker;
 import java.util.Optional;
-import org.junit.BeforeClass;
 import org.junit.Test;
 
 public class TestUdfWithNoPackage {
 
   private final UdfCompiler udfCompiler = new UdfCompiler(Optional.empty());
 
-  @BeforeClass
-  public static void ensureTestHasNoPackage() throws ClassNotFoundException {
-    // guard against someone accidentally refactoring and moving this test
-    // into a non-empty package
-    assertThat(Class.forName("TestUdfWithNoPackage").getPackage().getName(), is(""));
-  }
-
   @Test
   public void shouldCompileMethodsWithNoPackage() throws Exception {
+    // Given:
+    double version = Double.parseDouble(System.getProperty("java.specification.version"));
+    if (version < 1.9) {
+      assertThat(this.getClass().getPackage(), nullValue());
+    } else {
+      assertThat(this.getClass().getPackage().getName(), is(""));
+    }
+
+    // When:
     // motivated by https://github.com/square/javapoet/pull/723
     final UdfInvoker udf = udfCompiler
         .compile(getClass().getMethod("udf"), this.getClass().getClassLoader());
 
+    // Then:
     assertThat(udf.eval(this), is("udf"));
   }
 

--- a/ksql-engine/src/test/java/TestUdfWithNoPackage.java
+++ b/ksql-engine/src/test/java/TestUdfWithNoPackage.java
@@ -13,13 +13,12 @@
  * specific language governing permissions and limitations under the License.
  */
 
-import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
 
 import io.confluent.ksql.function.UdfCompiler;
 import io.confluent.ksql.function.UdfInvoker;
 import java.util.Optional;
-import org.junit.Before;
 import org.junit.BeforeClass;
 import org.junit.Test;
 
@@ -28,10 +27,10 @@ public class TestUdfWithNoPackage {
   private final UdfCompiler udfCompiler = new UdfCompiler(Optional.empty());
 
   @BeforeClass
-  public static void ensureTestHasNoPackage() {
+  public static void ensureTestHasNoPackage() throws ClassNotFoundException {
     // guard against someone accidentally refactoring and moving this test
     // into a non-empty package
-    assertThat(TestUdfWithNoPackage.class.getPackage().getName(), is(""));
+    assertThat(Class.forName("TestUdfWithNoPackage").getPackage().getName(), is(""));
   }
 
   @Test

--- a/ksql-engine/src/test/java/TestUdfWithNoPackage.java
+++ b/ksql-engine/src/test/java/TestUdfWithNoPackage.java
@@ -1,0 +1,40 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.CoreMatchers.is;
+
+import io.confluent.ksql.function.UdfCompiler;
+import io.confluent.ksql.function.UdfInvoker;
+import java.util.Optional;
+import org.junit.Test;
+
+public class TestUdfWithNoPackage {
+
+  private final UdfCompiler udfCompiler = new UdfCompiler(Optional.empty());
+
+  @Test
+  public void shouldCompileMethodsWithNoPackage() throws Exception {
+    final UdfInvoker udf = udfCompiler
+        .compile(getClass().getMethod("udf"), this.getClass().getClassLoader());
+
+    assertThat(udf.eval(this), is("udf"));
+  }
+
+  public String udf() {
+    return "udf";
+  }
+
+}

